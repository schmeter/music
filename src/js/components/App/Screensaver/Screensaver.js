import React from 'react';
import PropTypes from 'prop-types';
import autobind from 'autobind-decorator';

import { isTouch } from '../../../util/screen';

import configApp from '../../../../config/app.json';


class Screensaver extends React.Component {
    timer = null;

    componentDidMount() {
        document.addEventListener('keydown', this.captureKeys);

        document.addEventListener('mousemove', () => {
            this.timer = this.resetScreenSaver(this.timer);
        });
    }

    @autobind
    captureKeys(e) {
        switch (e.keyCode) {
            case 37:
                break;
            case 39:
                break;
            default:
                this.timer = this.resetScreenSaver(this.timer);
        }
    }

    screenSaverAllowed() {
        const { isPlaying } = this.props;
        return !isTouch() && configApp.useScreensaver && isPlaying;
    }

    startScreenSaver() {
        if (this.screenSaverAllowed()) {
            document.body.classList.add('screensaver');
        }
    }

    resetScreenSaver(timer) {
        document.body.classList.remove('screensaver');
        clearTimeout(timer);
        return setTimeout(() => {
            this.startScreenSaver();
        }, 3 * 10000);
    }

    render() {
        return (
            <div className="screensaver" />
        );
    }
}

Screensaver.propTypes = {
    isPlaying: PropTypes.bool.isRequired
};

export default Screensaver;
