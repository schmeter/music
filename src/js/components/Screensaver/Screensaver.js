import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import { isTouch } from '../../helpers/screen';

class Screensaver extends React.Component {
    state = {
      active: false,
    };

    timer = null;

    componentDidMount() {
      document.addEventListener('keydown', this.captureKeys);
      document.addEventListener('mousemove', () => {
        this.timer = this.resetScreenSaver(this.timer);
      });
    }

    captureKeys = e => {
      switch (e.keyCode) {
        case 37:
          break;
        case 39:
          break;
        default:
          this.timer = this.resetScreenSaver(this.timer);
      }
    }

    screenSaverAllowed() {
      const { config, isPlaying } = this.props;

      return !isTouch() && config.active && isPlaying;
    }

    startScreenSaver() {
      if (this.screenSaverAllowed()) {
        this.setState({
          active: true,
        });
      }
    }

    resetScreenSaver(timer) {
      const { config } = this.props;

      this.setState({
        active: false,
      });
      clearTimeout(timer);
      return setTimeout(() => {
        this.startScreenSaver();
      }, config.timeout * 1000);
    }

    render() {
      const { children } = this.props;
      const { active } = this.state;

      return (
        <div
          className={classNames(
            'screensaver',
            active && 'active',
          )}
        >
          {children}
        </div>
      );
    }
}

Screensaver.propTypes = {
  config: PropTypes.shape({
    timeout: PropTypes.number.isRequired,
    active: PropTypes.bool.isRequired,
  }).isRequired,
  children: PropTypes.node.isRequired,
  isPlaying: PropTypes.bool.isRequired,
};

export default Screensaver;
