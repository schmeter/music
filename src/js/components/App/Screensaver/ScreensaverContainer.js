import { connect } from 'react-redux';

import Screensaver from './Screensaver';
import { isPlaying } from '../../../state/selectors/audio';

export const mapStateToProps = state => ({
    isPlaying: isPlaying(state),
});

export default connect(mapStateToProps, null)(Screensaver);
