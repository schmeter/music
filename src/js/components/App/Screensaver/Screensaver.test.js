import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Screensaver from './Screensaver';

jest.mock('../../../util/screen', () => ({
    isTouch: () => false
}));

window.setTimeout = (callback, value) => {
    callback();
    return value;
};

it('renders correctly', () => {
    let props = {
        isPlaying: false
    };

    let component = shallow(<Screensaver {...props} />);

    expect(toJson(component)).toMatchSnapshot();

    props = {
        ...props,
        isPlaying: true
    };

    component.setProps(props);
    component.update();
    expect(toJson(component)).toMatchSnapshot();

    document.dispatchEvent(new window.KeyboardEvent('keydown', { keyCode: 0 }));
    document.dispatchEvent(new window.KeyboardEvent('keydown', { keyCode: 37 }));
    document.dispatchEvent(new window.KeyboardEvent('keydown', { keyCode: 39 }));

    document.dispatchEvent(new window.KeyboardEvent('mousemove', {}));

    props = {
        ...props,
        isPlaying: false
    };

    component.setProps(props);
    component.update();
    expect(toJson(component)).toMatchSnapshot();

    document.dispatchEvent(new window.KeyboardEvent('mousemove', {}));
});
