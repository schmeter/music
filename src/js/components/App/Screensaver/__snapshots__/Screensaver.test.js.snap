// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="screensaver"
/>
`;

exports[`renders correctly 2`] = `
<div
  className="screensaver"
/>
`;

exports[`renders correctly 3`] = `
<div
  className="screensaver"
/>
`;
